agent/agent.go

```diff
@@ -281,59 +281,55 @@ func (a *Agent) runInputs(
 ) {
 	var wg sync.WaitGroup
 	for _, input := range unit.inputs {
-		a.runInput(ctx, startTime, unit, input, &wg)
-	}
+		// Overwrite agent interval if this plugin has its own.
+		interval := time.Duration(a.Config.Agent.Interval)
+		if input.Config.Interval != 0 {
+			interval = input.Config.Interval
+		}
 
-	wg.Wait()
+		// Overwrite agent precision if this plugin has its own.
+		precision := time.Duration(a.Config.Agent.Precision)
+		if input.Config.Precision != 0 {
+			precision = input.Config.Precision
+		}
 
-	log.Printf("D! [agent] Stopping service inputs")
-	stopServiceInputs(unit.inputs)
+		// Overwrite agent collection_jitter if this plugin has its own.
+		jitter := time.Duration(a.Config.Agent.CollectionJitter)
+		if input.Config.CollectionJitter != 0 {
+			jitter = input.Config.CollectionJitter
+		}
 
-	close(unit.dst)
-	log.Printf("D! [agent] Input channel closed")
-}
+		// Overwrite agent collection_offset if this plugin has its own.
+		offset := time.Duration(a.Config.Agent.CollectionOffset)
+		if input.Config.CollectionOffset != 0 {
+			offset = input.Config.CollectionOffset
+		}
 
-func (a *Agent) runInput(ctx context.Context, startTime time.Time, unit *inputUnit, input *models.RunningInput, wg *sync.WaitGroup) {
-	// Overwrite agent interval if this plugin has its own.
-	interval := time.Duration(a.Config.Agent.Interval)
-	if input.Config.Interval != 0 {
-		interval = input.Config.Interval
-	}
+		var ticker Ticker
+		if a.Config.Agent.RoundInterval {
+			ticker = NewAlignedTicker(startTime, interval, jitter, offset)
+		} else {
+			ticker = NewUnalignedTicker(interval, jitter, offset)
+		}
+		defer ticker.Stop()
 
-	// Overwrite agent precision if this plugin has its own.
-	precision := time.Duration(a.Config.Agent.Precision)
-	if input.Config.Precision != 0 {
-		precision = input.Config.Precision
-	}
+		acc := NewAccumulator(input, unit.dst)
+		acc.SetPrecision(getPrecision(precision, interval))
 
-	// Overwrite agent collection_jitter if this plugin has its own.
-	jitter := time.Duration(a.Config.Agent.CollectionJitter)
-	if input.Config.CollectionJitter != 0 {
-		jitter = input.Config.CollectionJitter
-	}
-
-	// Overwrite agent collection_offset if this plugin has its own.
-	offset := time.Duration(a.Config.Agent.CollectionOffset)
-	if input.Config.CollectionOffset != 0 {
-		offset = input.Config.CollectionOffset
+		wg.Add(1)
+		go func(input *models.RunningInput) {
+			defer wg.Done()
+			a.gatherLoop(ctx, acc, input, ticker, interval)
+		}(input)
 	}
 
-	var ticker Ticker
-	if a.Config.Agent.RoundInterval {
-		ticker = NewAlignedTicker(startTime, interval, jitter, offset)
-	} else {
-		ticker = NewUnalignedTicker(interval, jitter, offset)
-	}
-	defer ticker.Stop()
```

Comment:
I made this change: https://github.com/influxdata/telegraf/commit/85b7590ff9c5a1abbf83aad034b8b7a4be3daf2e#diff-62b6ad581fe3a3059ae8c85ef0f31dde4092bfdecfa7d6857c470bcacaa8cc8bL308-L332 to avoid possible resource leak (`defer` is called in the `for` loop). 

But it seems that thanks to this faulty(?) behavior Telegraf works 😄

We can revert this function but we need to think how to prevent `defer ticker.Stop()` from leaking and have Telegraf which works ;)

Reply:
I just pushed what I think was missing. I don't know my wait groups enough, but this does seem to restore the correct behavior.

Reply:
I don't think it is the problem with `wg`.

After my changes, `ticker.Stop()` is called just after `runInput` function is finished. Thanks to that, plugin will not be able to `Gather` at even once.
Before `ticker.Stop()` was called just after `runInputs` function was finished (== when Telegraf process was going to finish).

Reply:
Ah thanks for digging in.